src/Security/AuthenticationMiddleware.php

```diff
@@ -46,9 +103,27 @@ public function setAuthenticationHandler(AuthenticationHandler $authenticationHa
     public function process(HTTPRequest $request, callable $delegate)
     {
         try {
-            $this
-                ->getAuthenticationHandler()
-                ->authenticateRequest($request);
+            $handler = $this->getAuthenticationHandler();
+            $handler->authenticateRequest($request);
+
+            if (static::config()->get('performExpiredPasswordCheck')) {
+                $session = $request->getSession();
+                if ($session && $session->get(static::SESSION_KEY_EXPIRED_PASSWORD_ALLOW_REQUEST)) {
+                    // allow current request and skip the expiration check, but for only the current
+                    // request, so we're deleting the flag from the session so it's not affecting other
+                    // requests.
+                    // This flag would usually be set from within $handler->authenticateRequest()
+
+                    $session->clear(static::SESSION_KEY_EXPIRED_PASSWORD_ALLOW_REQUEST);
+                } else {
+                    $user = Security::getCurrentUser();
+                    if ($user && $user->isPasswordExpired()) {
+                        if ($response = $this->handleExpiredPassword($request)) {
+                            return $response;
+                        }
+                    }
+                }
```

Comment:
I'm leaning towards suggesting that this could be its own `ForceChangePasswordMiddleware` or something. Thoughts?

Reply:
yes, thought about it as well, since the original 63 lines of code became 263. However, then I had a look at what AuthenticationMiddleware actually does and there's not much left for it to do anyway. Checking for the expired passwords feels to be in authentication domain to me anyway.

Reply:
Perhaps. Visually this method doesn't look particularly nice to me, so I think if you leave it here you may be able to do a little refactor to tidy it up a bit, e.g. combine conditions, maybe move it into its own method or something. I'm OK with leaving it in this class if you think it belongs here.